server/src/main/java/org/opensearch/indices/tiering/HotToWarmTieringService.java

```diff
@@ -0,0 +1,875 @@
+    protected final IndexNameExpressionResolver indexNameExpressionResolver;
+    protected final AllocationService allocationService;
+    // map to store all the in progress tiering requests
+    private final Map<String, TieringRequestContext> requestUuidToRequestContext = new ConcurrentHashMap<>();
+    // map to store index info of all the indices under-going hot to warm tiering
+    private final Map<Index, IndexTieringInfo> indexTieringInfoMap = new ConcurrentHashMap<>();
+    public static final String HOT_TO_WARM_START_TIME = "hot_to_warm_start_time";
+    public static final String HOT_TO_WARM_END_TIME = "hot_to_warm_end_time";
+
+    @Inject
+    public HotToWarmTieringService(
+        Settings settings,
+        ClusterService clusterService,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        AllocationService allocationService
+    ) {
+        super();
+        this.clusterService = clusterService;
+        this.indexNameExpressionResolver = indexNameExpressionResolver;
+        this.allocationService = allocationService;
+
+        if (DiscoveryNode.isClusterManagerNode(settings) && FeatureFlags.isEnabled(FeatureFlags.TIERED_REMOTE_INDEX)) {
+            clusterService.addListener(this);
+        }
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        // TODO add handling for master switch, dangling indices, master reload
+        if (event.routingTableChanged()) {
+            if (!indexTieringInfoMap.isEmpty()) {
+                updateIndexShardStatus(event.state());
+            }
+            if (!requestUuidToRequestContext.isEmpty()) {
+                completeRequestLevelTiering();
+            }
+        }
+    }
+
+    /**
+     * Checks and completes the request level tiering for
+     * the request uuids in requestUuidToRequestContext map.
+     */
+    void completeRequestLevelTiering() {
+        for (String requestUuid : requestUuidToRequestContext.keySet()) {
+            final TieringRequestContext requestContext = requestUuidToRequestContext.get(requestUuid);
+            if (requestContext.isRequestProcessingComplete()) {
+                logger.info("Tiering is completed for the request [{}]", requestContext);
+                for (Index index : indexTieringInfoMap.keySet()) {
+                    indexTieringInfoMap.remove(index);
+                }
+                requestUuidToRequestContext.remove(requestUuid);
+                if (requestContext.getRequest().waitForCompletion()) {
+                    requestContext.getListener().onResponse(requestContext.constructHotToWarmTieringResponse());
+                }
+            }
+        }
+    }
+
+    /**
+     * Checks if the given node is a search node.
+     * @param node the node to check.
+     * @return true if the node is a search node, false otherwise.
+     */
+    boolean isSearchNode(final DiscoveryNode node) {
+        return node != null && node.isSearchNode();
+    }
```

Comment:
should we add this null check in isSearchNode() method of DiscoveryNode class?

Reply:
this check is already there in DiscoveryNode - this check adds a null check also

Reply:
i removed this to avoid confusion